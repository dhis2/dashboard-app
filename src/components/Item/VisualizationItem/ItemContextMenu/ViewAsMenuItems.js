import i18n from '@dhis2/d2-i18n'
import { IconVisualizationColumn16, IconTable16, IconWorld16 } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React from 'react'
import {
    CHART,
    MAP,
    REPORT_TABLE,
    EVENT_CHART,
    EVENT_REPORT,
    EVENT_VISUALIZATION,
    isTrackerDomainType,
    hasMapView,
} from '../../../../modules/itemTypes.js'
import MenuItem from '../../../MenuItemWithTooltip.js'
import getThematicMapViews from '../getThematicMapViews.js'

const ViewAsMenuItems = ({
    type,
    activeType,
    visualization,
    onActiveTypeChanged,
}) => {
    const isTrackerType = isTrackerDomainType(type)

    const onViewTable = () =>
        onActiveTypeChanged(isTrackerType ? EVENT_REPORT : REPORT_TABLE)

    const onViewChart = () =>
        onActiveTypeChanged(isTrackerType ? EVENT_CHART : CHART)

    const onViewMap = () => onActiveTypeChanged(MAP)

    const notSupported = type === MAP && !getThematicMapViews(visualization)

    return (
        <>
            {![CHART, EVENT_CHART].includes(activeType) && (
                <MenuItem
                    tooltip={
                        notSupported
                            ? i18n.t("This map can't be displayed as a chart")
                            : null
                    }
                    label={i18n.t('View as Chart')}
                    onClick={onViewChart}
                    disabled={notSupported}
                    icon={<IconVisualizationColumn16 />}
                />
            )}
            {![REPORT_TABLE, EVENT_REPORT, EVENT_VISUALIZATION].includes(
                activeType
            ) && (
                <MenuItem
                    tooltip={
                        notSupported
                            ? i18n.t("This map can't be displayed as a table")
                            : null
                    }
                    label={i18n.t('View as Table')}
                    onClick={onViewTable}
                    disabled={notSupported}
                    icon={<IconTable16 />}
                />
            )}
            {hasMapView(type) && activeType !== MAP && (
                <MenuItem
                    label={i18n.t('View as Map')}
                    onClick={onViewMap}
                    icon={<IconWorld16 />}
                />
            )}
        </>
    )
}

ViewAsMenuItems.propTypes = {
    activeType: PropTypes.string,
    type: PropTypes.string,
    visualization: PropTypes.object,
    onActiveTypeChanged: PropTypes.func,
}

export default ViewAsMenuItems
