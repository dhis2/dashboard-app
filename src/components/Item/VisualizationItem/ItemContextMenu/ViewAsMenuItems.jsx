import { VIS_TYPE_OUTLIER_TABLE } from '@dhis2/analytics'
import i18n from '@dhis2/d2-i18n'
import { IconVisualizationColumn16, IconTable16, IconWorld16 } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React from 'react'
import {
    CHART,
    MAP,
    REPORT_TABLE,
    EVENT_CHART,
    EVENT_REPORT,
    EVENT_VISUALIZATION,
    isTrackerDomainType,
    hasMapView,
} from '../../../../modules/itemTypes.js'
import MenuItem from '../../../MenuItemWithTooltip.jsx'
import getThematicMapViews from '../getThematicMapViews.js'

const ViewAsMenuItems = ({
    type,
    activeType,
    visualization,
    onActiveTypeChanged,
}) => {
    const isTrackerType = isTrackerDomainType(type)

    const onViewTable = () =>
        onActiveTypeChanged(isTrackerType ? EVENT_REPORT : REPORT_TABLE)

    const onViewChart = () =>
        onActiveTypeChanged(isTrackerType ? EVENT_CHART : CHART)

    const onViewMap = () => onActiveTypeChanged(MAP)

    const notSupported =
        (type === MAP && !getThematicMapViews(visualization)) ||
        (type === CHART && visualization.type === VIS_TYPE_OUTLIER_TABLE)

    const getNotSupportedMessage = (viewAs) => {
        if (type === MAP && !getThematicMapViews(visualization)) {
            return viewAs === 'chart'
                ? i18n.t("This map can't be displayed as a chart")
                : i18n.t("This map can't be displayed as a pivot table")
        }

        if (type === CHART && visualization.type === VIS_TYPE_OUTLIER_TABLE) {
            return viewAs === 'table'
                ? i18n.t(
                      "This visualization can't be displayed as a pivot table"
                  )
                : i18n.t("This visualization can't be displayed as a map")
        }

        return null
    }

    return (
        <>
            {![CHART, EVENT_CHART].includes(activeType) && (
                <MenuItem
                    tooltip={getNotSupportedMessage('chart')}
                    label={i18n.t('View as Chart')}
                    onClick={onViewChart}
                    disabled={notSupported}
                    icon={<IconVisualizationColumn16 />}
                />
            )}
            {![REPORT_TABLE, EVENT_REPORT, EVENT_VISUALIZATION].includes(
                activeType
            ) && (
                <MenuItem
                    tooltip={getNotSupportedMessage('table')}
                    label={i18n.t('View as Pivot table')}
                    onClick={onViewTable}
                    disabled={notSupported}
                    icon={<IconTable16 />}
                />
            )}
            {hasMapView(type) && activeType !== MAP && (
                <MenuItem
                    tooltip={getNotSupportedMessage('map')}
                    label={i18n.t('View as Map')}
                    onClick={onViewMap}
                    disabled={notSupported}
                    icon={<IconWorld16 />}
                />
            )}
        </>
    )
}

ViewAsMenuItems.propTypes = {
    activeType: PropTypes.string,
    type: PropTypes.string,
    visualization: PropTypes.object,
    onActiveTypeChanged: PropTypes.func,
}

export default ViewAsMenuItems
