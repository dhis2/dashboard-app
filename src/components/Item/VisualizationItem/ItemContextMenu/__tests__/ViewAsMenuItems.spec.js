import React from 'react'
import { render } from '@testing-library/react'
import ViewAsMenuItems from '../ViewAsMenuItems'
import {
    REPORT_TABLE,
    MAP,
    CHART,
    EVENT_REPORT,
    EVENT_CHART,
} from '../../../../../modules/itemTypes'

const defaultProps = {
    onActiveTypeChanged: jest.fn(),
}

test('renders menu for active type MAP and type CHART', async () => {
    const props = Object.assign({}, defaultProps, {
        type: CHART,
        activeType: MAP,
    })

    const { container } = render(<ViewAsMenuItems {...props} />)

    expect(container).toMatchSnapshot()
})

test('renders menu for active type CHART and type MAP', async () => {
    const props = Object.assign({}, defaultProps, {
        type: MAP,
        activeType: CHART,
    })

    const { container } = render(<ViewAsMenuItems {...props} />)

    expect(container).toMatchSnapshot()
})

test('renders menu for active type REPORT_TABLE and type CHART', async () => {
    const props = Object.assign({}, defaultProps, {
        type: CHART,
        activeType: REPORT_TABLE,
    })

    const { container } = render(<ViewAsMenuItems {...props} />)

    expect(container).toMatchSnapshot()
})

test('renders menu for active type CHART and type REPORT_TABLE', async () => {
    const props = Object.assign({}, defaultProps, {
        type: REPORT_TABLE,
        activeType: CHART,
    })

    const { container } = render(<ViewAsMenuItems {...props} />)

    expect(container).toMatchSnapshot()
})

test('renders menu for active type EVENT_REPORT and type EVENT_CHART', async () => {
    const props = Object.assign({}, defaultProps, {
        type: EVENT_CHART,
        activeType: EVENT_REPORT,
    })

    const { container } = render(<ViewAsMenuItems {...props} />)

    expect(container).toMatchSnapshot()
})

test('renders menu for active type EVENT_CHART and type EVENT_REPORT', async () => {
    const props = Object.assign({}, defaultProps, {
        type: EVENT_REPORT,
        activeType: EVENT_CHART,
    })

    const { container } = render(<ViewAsMenuItems {...props} />)

    expect(container).toMatchSnapshot()
})
