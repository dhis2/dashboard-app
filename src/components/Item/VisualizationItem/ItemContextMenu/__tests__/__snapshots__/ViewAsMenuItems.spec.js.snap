// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders menu for active type CHART and type MAP 1`] = `
<div>
  <li
    class="jsx-665727467 dense"
    data-test="dhis2-uicore-menuitem"
  >
    <a
      class="jsx-665727467"
    >
      <span
        class="jsx-665727467 icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-10"
          focusable="false"
          role="presentation"
          style="fill: #6e7a8a;"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 14h4v-4H4v4zm0 5h4v-4H4v4zM4 9h4V5H4v4zm5 5h12v-4H9v4zm0 5h12v-4H9v4zM9 5v4h12V5H9z"
          />
        </svg>
      </span>
      <span
        class="jsx-665727467 label"
      >
        View as Table
      </span>
    </a>
  </li>
  <li
    class="jsx-665727467 dense"
    data-test="dhis2-uicore-menuitem"
  >
    <a
      class="jsx-665727467"
    >
      <span
        class="jsx-665727467 icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-10"
          focusable="false"
          role="presentation"
          style="fill: #6e7a8a;"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-1 17.93c-3.95-.49-7-3.85-7-7.93 0-.62.08-1.21.21-1.79L9 15v1c0 1.1.9 2 2 2v1.93zm6.9-2.54c-.26-.81-1-1.39-1.9-1.39h-1v-3c0-.55-.45-1-1-1H8v-2h2c.55 0 1-.45 1-1V7h2c1.1 0 2-.9 2-2v-.41c2.93 1.19 5 4.06 5 7.41 0 2.08-.8 3.97-2.1 5.39z"
          />
        </svg>
      </span>
      <span
        class="jsx-665727467 label"
      >
        View as Map
      </span>
    </a>
  </li>
</div>
`;

exports[`renders menu for active type CHART and type REPORT_TABLE 1`] = `
<div>
  <li
    class="jsx-665727467 dense"
    data-test="dhis2-uicore-menuitem"
  >
    <a
      class="jsx-665727467"
    >
      <span
        class="jsx-665727467 icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-28"
          focusable="false"
          role="presentation"
          style="fill: #6e7a8a;"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 14h4v-4H4v4zm0 5h4v-4H4v4zM4 9h4V5H4v4zm5 5h12v-4H9v4zm0 5h12v-4H9v4zM9 5v4h12V5H9z"
          />
        </svg>
      </span>
      <span
        class="jsx-665727467 label"
      >
        View as Table
      </span>
    </a>
  </li>
  <li
    class="jsx-665727467 dense"
    data-test="dhis2-uicore-menuitem"
  >
    <a
      class="jsx-665727467"
    >
      <span
        class="jsx-665727467 icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-28"
          focusable="false"
          role="presentation"
          style="fill: #6e7a8a;"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-1 17.93c-3.95-.49-7-3.85-7-7.93 0-.62.08-1.21.21-1.79L9 15v1c0 1.1.9 2 2 2v1.93zm6.9-2.54c-.26-.81-1-1.39-1.9-1.39h-1v-3c0-.55-.45-1-1-1H8v-2h2c.55 0 1-.45 1-1V7h2c1.1 0 2-.9 2-2v-.41c2.93 1.19 5 4.06 5 7.41 0 2.08-.8 3.97-2.1 5.39z"
          />
        </svg>
      </span>
      <span
        class="jsx-665727467 label"
      >
        View as Map
      </span>
    </a>
  </li>
</div>
`;

exports[`renders menu for active type EVENT_CHART and type EVENT_REPORT 1`] = `
<div>
  <li
    class="jsx-665727467 dense"
    data-test="dhis2-uicore-menuitem"
  >
    <a
      class="jsx-665727467"
    >
      <span
        class="jsx-665727467 icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-46"
          focusable="false"
          role="presentation"
          style="fill: #6e7a8a;"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 14h4v-4H4v4zm0 5h4v-4H4v4zM4 9h4V5H4v4zm5 5h12v-4H9v4zm0 5h12v-4H9v4zM9 5v4h12V5H9z"
          />
        </svg>
      </span>
      <span
        class="jsx-665727467 label"
      >
        View as Table
      </span>
    </a>
  </li>
</div>
`;

exports[`renders menu for active type EVENT_REPORT and type EVENT_CHART 1`] = `
<div>
  <li
    class="jsx-665727467 dense"
    data-test="dhis2-uicore-menuitem"
  >
    <a
      class="jsx-665727467"
    >
      <span
        class="jsx-665727467 icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-37"
          focusable="false"
          role="presentation"
          style="fill: #6e7a8a;"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"
          />
        </svg>
      </span>
      <span
        class="jsx-665727467 label"
      >
        View as Chart
      </span>
    </a>
  </li>
</div>
`;

exports[`renders menu for active type MAP and type CHART 1`] = `
<div>
  <li
    class="jsx-665727467 dense"
    data-test="dhis2-uicore-menuitem"
  >
    <a
      class="jsx-665727467"
    >
      <span
        class="jsx-665727467 icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-1"
          focusable="false"
          role="presentation"
          style="fill: #6e7a8a;"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"
          />
        </svg>
      </span>
      <span
        class="jsx-665727467 label"
      >
        View as Chart
      </span>
    </a>
  </li>
  <li
    class="jsx-665727467 dense"
    data-test="dhis2-uicore-menuitem"
  >
    <a
      class="jsx-665727467"
    >
      <span
        class="jsx-665727467 icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-1"
          focusable="false"
          role="presentation"
          style="fill: #6e7a8a;"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 14h4v-4H4v4zm0 5h4v-4H4v4zM4 9h4V5H4v4zm5 5h12v-4H9v4zm0 5h12v-4H9v4zM9 5v4h12V5H9z"
          />
        </svg>
      </span>
      <span
        class="jsx-665727467 label"
      >
        View as Table
      </span>
    </a>
  </li>
</div>
`;

exports[`renders menu for active type REPORT_TABLE and type CHART 1`] = `
<div>
  <li
    class="jsx-665727467 dense"
    data-test="dhis2-uicore-menuitem"
  >
    <a
      class="jsx-665727467"
    >
      <span
        class="jsx-665727467 icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-19"
          focusable="false"
          role="presentation"
          style="fill: #6e7a8a;"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"
          />
        </svg>
      </span>
      <span
        class="jsx-665727467 label"
      >
        View as Chart
      </span>
    </a>
  </li>
  <li
    class="jsx-665727467 dense"
    data-test="dhis2-uicore-menuitem"
  >
    <a
      class="jsx-665727467"
    >
      <span
        class="jsx-665727467 icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-19"
          focusable="false"
          role="presentation"
          style="fill: #6e7a8a;"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-1 17.93c-3.95-.49-7-3.85-7-7.93 0-.62.08-1.21.21-1.79L9 15v1c0 1.1.9 2 2 2v1.93zm6.9-2.54c-.26-.81-1-1.39-1.9-1.39h-1v-3c0-.55-.45-1-1-1H8v-2h2c.55 0 1-.45 1-1V7h2c1.1 0 2-.9 2-2v-.41c2.93 1.19 5 4.06 5 7.41 0 2.08-.8 3.97-2.1 5.39z"
          />
        </svg>
      </span>
      <span
        class="jsx-665727467 label"
      >
        View as Map
      </span>
    </a>
  </li>
</div>
`;
